In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC, LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn import metrics
from sklearn import preprocessing

In [ ]:
df = pd.read_csv('/content/emails.csv')

In [ ]:
df.info

<bound method DataFrame.info of        Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  \
0        Email 1    0   0    1    0    0   0    2    0    0  ...         0   
1        Email 2    8  13   24    6    6   2  102    1   27  ...         0   
2        Email 3    0   0    1    0    0   0    8    0    0  ...         0   
3        Email 4    0   5   22    0    5   1   51    2   10  ...         0   
4        Email 5    7   6   17    1    5   2   57    0    9  ...         0   
...          ...  ...  ..  ...  ...  ...  ..  ...  ...  ...  ...       ...   
5167  Email 5168    2   2    2    3    0   0   32    0    0  ...         0   
5168  Email 5169   35  27   11    2    6   5  151    4    3  ...         0   
5169  Email 5170    0   0    1    1    0   0   11    0    0  ...         0   
5170  Email 5171    2   7    1    0    2   1   28    2    0  ...         0   
5171  Email 5172   22  24    5    1    6   5  148    8    2  ...         0   

      jay  valued  lay  infrast

In [ ]:
df.isnull().any()

Email No.     False
the           False
to            False
ect           False
and           False
              ...  
military      False
allowing      False
ff            False
dry           False
Prediction    False
Length: 3002, dtype: bool

In [ ]:
df.drop(columns='Email No.', inplace=True)
df

the  to  ect  and  for  of    a  you  hou  in  ...  connevey  jay  \
0       0   0    1    0    0   0    2    0    0   0  ...         0    0   
1       8  13   24    6    6   2  102    1   27  18  ...         0    0   
2       0   0    1    0    0   0    8    0    0   4  ...         0    0   
3       0   5   22    0    5   1   51    2   10   1  ...         0    0   
4       7   6   17    1    5   2   57    0    9   3  ...         0    0   
...   ...  ..  ...  ...  ...  ..  ...  ...  ...  ..  ...       ...  ...   
5167    2   2    2    3    0   0   32    0    0   5  ...         0    0   
5168   35  27   11    2    6   5  151    4    3  23  ...         0    0   
5169    0   0    1    1    0   0   11    0    0   1  ...         0    0   
5170    2   7    1    0    2   1   28    2    0   8  ...         0    0   
5171   22  24    5    1    6   5  148    8    2  23  ...         0    0   

      valued  lay  infrastructure  military  allowing  ff  dry  Prediction  
0          0    0               0         0         0   0    0           0  
1          0    0               0         0         0   1    0           0  
2          0    0               0         0         0   0    0           0  
3          0    0               0         0         0   0    0           0  
4          0    0               0         0         0   1    0           0  
...      ...  ...             ...       ...       ...  ..  ...         ...  
5167       0    0               0         0         0   0    0           0  
5168       0    0               0         0         0   1    0           0  
5169       0    0               0         0         0   0    0           1  
5170       0    0               0         0         0   1    0           1  
5171       0    0               0         0         0   0    0           0  

[5172 rows x 3001 columns]

In [ ]:
df.dtypes

the           int64
to            int64
ect           int64
and           int64
for           int64
              ...  
military      int64
allowing      int64
ff            int64
dry           int64
Prediction    int64
Length: 3001, dtype: object

In [ ]:
df.columns

Index(['the', 'to', 'ect', 'and', 'for', 'of', 'a', 'you', 'hou', 'in',
       ...
       'connevey', 'jay', 'valued', 'lay', 'infrastructure', 'military',
       'allowing', 'ff', 'dry', 'Prediction'],
      dtype='object', length=3001)

In [ ]:
df.Prediction.unique()

array([0, 1], dtype=int64)

In [ ]:
df['Prediction'] = df['Prediction'].replace({0:'Not spam', 1:'Spam'})
df['Prediction']

0       Not spam
1       Not spam
2       Not spam
3       Not spam
4       Not spam
          ...   
5167    Not spam
5168    Not spam
5169        Spam
5170        Spam
5171    Not spam
Name: Prediction, Length: 5172, dtype: object

# Machine Learning models
The following 5 models are used:

1. K-Nearest Neighbors
2. Linear SVM
3. Polynomial SVM
4. RBF SVM
5. Sigmoid SVM

In [ ]:
#KNN

X = df.drop(columns='Prediction',axis = 1)
Y = df['Prediction']

X.columns

Index(['the', 'to', 'ect', 'and', 'for', 'of', 'a', 'you', 'hou', 'in',
       ...
       'enhancements', 'connevey', 'jay', 'valued', 'lay', 'infrastructure',
       'military', 'allowing', 'ff', 'dry'],
      dtype='object', length=3000)

In [ ]:
Y.head()

0    Not spam
1    Not spam
2    Not spam
3    Not spam
4    Not spam
Name: Prediction, dtype: object

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=1)

In [ ]:
KN = KNeighborsClassifier
knn = KN(n_neighbors=7)
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)

In [ ]:
print("Prediction: \n")
print(y_pred)

Prediction: 

['Not spam' 'Spam' 'Not spam' ... 'Not spam' 'Not spam' 'Not spam']


In [ ]:
# Accuracy of KNN

M = metrics.accuracy_score(y_test,y_pred)
print("KNN accuracy: ", M)

KNN accuracy:  0.8714975845410629


In [ ]:
#confusion matrix
cm = metrics.confusion_matrix(y_test,y_pred)
print("Confusion matrix: \n", cm)

Confusion matrix: 
 [[635  84]
 [ 49 267]]


In [ ]:
#SVM Classifier

svc = SVC(C=1.0,kernel='rbf',gamma='auto')
# C here is the regularization parameter. Here, L2 penalty is used(default). It is the inverse of the strength of regularization.
# As C increases, model overfits.
# Kernel here is the radial basis function kernel.
# gamma (only used for rbf kernel) : As gamma increases, model overfits.
svc.fit(x_train, y_train)
y_pred = svc.predict(x_test)
print("Accuracy Score for SVC : ", accuracy_score(y_pred,y_test))

Accuracy Score for SVC :  0.9014492753623189


In [ ]:
#confusion matrix
cm = metrics.confusion_matrix(y_test,y_pred)
print("Confusion matrix: \n", cm)

Confusion matrix: 
 [[690  29]
 [ 73 243]]


# **# Explanation:**


# **bold text**
 Certainly! Let's break down the code in detail, explaining each line and the overall process for both the K-Nearest Neighbors (KNN) and Support Vector Machine (SVM) classifiers.

### K-Nearest Neighbors (KNN) Classifier:

```python
X = df.drop(columns='Prediction', axis=1)
Y = df['Prediction']
```

- **Explanation:** Here, `X` is created to represent the features (independent variables), and `Y` is the target variable (dependent variable). `X` contains all the columns from the DataFrame `df` except the 'Prediction' column, and `Y` contains the 'Prediction' column.

```python
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=1)
```

- **Explanation:** The dataset is split into training and testing sets using the `train_test_split` function from scikit-learn. 80% of the data is used for training (`x_train`, `y_train`), and 20% is used for testing (`x_test`, `y_test`). The `random_state` parameter ensures reproducibility.

```python
KN = KNeighborsClassifier
knn = KN(n_neighbors=7)
```

- **Explanation:** The K-Nearest Neighbors classifier is imported (`KNeighborsClassifier`), and an instance of the KNN model is created with `n_neighbors` set to 7. This means it considers the 7 nearest neighbors for classification.

```python
knn.fit(x_train, y_train)
y_pred = knn.predict(x_test)
```

- **Explanation:** The KNN model is trained on the training data (`x_train`, `y_train`) using the `fit` method. Then, predictions are made on the testing data (`x_test`) using the `predict` method.

```python
M = metrics.accuracy_score(y_test, y_pred)
print("KNN accuracy: ", M)
```

- **Explanation:** The accuracy of the KNN model is calculated using the `accuracy_score` function from scikit-learn. Accuracy is the ratio of correctly predicted instances to the total instances in the testing set. The result is printed.

```python
cm = metrics.confusion_matrix(y_test, y_pred)
print("Confusion matrix: \n", cm)
```

- **Explanation:** The confusion matrix is calculated using the `confusion_matrix` function from scikit-learn. A confusion matrix provides information about the performance of a classification algorithm, showing the number of true positive, true negative, false positive, and false negative predictions.

### Support Vector Machine (SVM) Classifier:

```python
svc = SVC(C=1.0, kernel='rbf', gamma='auto')
```

- **Explanation:** The Support Vector Machine classifier is imported (`SVC`). An instance of the SVM model is created with parameters like `C` (regularization parameter), `kernel` (radial basis function kernel), and `gamma` (kernel coefficient).

```python
svc.fit(x_train, y_train)
y_pred = svc.predict(x_test)
```

- **Explanation:** The SVM model is trained on the training data (`x_train`, `y_train`) using the `fit` method. Predictions are then made on the testing data (`x_test`) using the `predict` method.

```python
print("Accuracy Score for SVC : ", accuracy_score(y_pred, y_test))
```

- **Explanation:** The accuracy of the SVM model is calculated using the `accuracy_score` function from scikit-learn and printed. Accuracy is the ratio of correctly predicted instances to the total instances in the testing set.

```python
cm = metrics.confusion_matrix(y_test, y_pred)
print("Confusion matrix: \n", cm)
```

- **Explanation:** Similar to the KNN section, the confusion matrix is calculated using the `confusion_matrix` function from scikit-learn. It provides insights into the performance of the SVM classifier.

### Summary:

- Both KNN and SVM are supervised learning algorithms used for classification tasks.
- The code demonstrates the typical workflow of loading data, splitting it into training and testing sets, training a model, making predictions, and evaluating the model's performance using accuracy and confusion matrix metrics.            Certainly! Let's go through the details, examples, and applications of each of the mentioned models:

### 1. K-Nearest Neighbors (KNN):

**Explanation:**
- **Model Overview:** K-Nearest Neighbors is a simple, instance-based learning algorithm for classification and regression. It classifies a data point based on the majority class of its k-nearest neighbors in the feature space.

**Example:**
- **Scenario:** Predicting whether an email is spam or not based on its word frequency.
- **Explanation:** If a new email has word frequencies similar to those of previously labeled spam emails, it is likely classified as spam.

**Applications:**
- **Recommendation Systems:** Recommending products or content based on user behavior.
- **Image Recognition:** Classifying an image based on similarities to other images in a dataset.
- **Anomaly Detection:** Identifying outliers or unusual patterns in data.

### 2. Linear Support Vector Machine (Linear SVM):

**Explanation:**
- **Model Overview:** Linear Support Vector Machine is a linear classifier that finds the hyperplane that best separates classes in the feature space. It aims to maximize the margin between classes.

**Example:**
- **Scenario:** Classifying emails as spam or not spam based on features like word frequency.
- **Explanation:** The hyperplane separates the feature space into regions for spam and non-spam emails.

**Applications:**
- **Text Classification:** Sentiment analysis, spam detection, and document categorization.
- **Image Classification:** Identifying objects or patterns in images.
- **Bioinformatics:** Classifying genes or proteins based on features.

### 3. Polynomial Support Vector Machine (Polynomial SVM):

**Explanation:**
- **Model Overview:** Polynomial Support Vector Machine is an extension of linear SVM that allows for non-linear decision boundaries by introducing polynomial kernels.

**Example:**
- **Scenario:** Classifying data points with a circular decision boundary.
- **Explanation:** The polynomial kernel introduces non-linearity, enabling the model to capture circular or curved patterns.

**Applications:**
- **Speech Recognition:** Identifying spoken words or phrases.
- **Computer Vision:** Recognizing objects in images with non-linear shapes.
- **Biomedical Engineering:** Classifying biological data with complex patterns.

### 4. Radial Basis Function (RBF) Support Vector Machine (SVM):

**Explanation:**
- **Model Overview:** RBF Support Vector Machine uses the radial basis function (Gaussian) kernel to handle complex, non-linear relationships in the feature space.

**Example:**
- **Scenario:** Separating data points in a dataset with intertwined classes.
- **Explanation:** The RBF kernel allows for more flexible decision boundaries, accommodating intricate relationships.

**Applications:**
- **Finance:** Predicting stock prices based on historical data.
- **Healthcare:** Diagnosing diseases based on patient data.
- **Chemistry:** Predicting molecular properties.

### 5. Sigmoid Support Vector Machine (Sigmoid SVM):

**Explanation:**
- **Model Overview:** Sigmoid Support Vector Machine uses the sigmoid function as a kernel to transform the data, enabling it to handle non-linear relationships.

**Example:**
- **Scenario:** Classifying data with a non-linear decision boundary.
- **Explanation:** The sigmoid kernel applies a logistic function to the input features, making it suitable for capturing complex patterns.

**Applications:**
- **Neural Network Activation Functions:** Sigmoid functions are commonly used as activation functions in neural networks.
- **Financial Fraud Detection:** Identifying fraudulent transactions based on patterns in financial data.
- **Bioinformatics:** Predicting biological phenomena based on complex data.

### Summary:

- **KNN** is a simple, versatile algorithm suitable for various tasks.
- **Linear SVM** is effective for linearly separable data.
- **Polynomial SVM** introduces non-linearity with polynomial kernels.
- **RBF SVM** handles complex relationships with Gaussian kernels.
- **Sigmoid SVM** transforms data using sigmoid functions for non-linear patterns.

Each model has its strengths and is suited for specific types of data and tasks. The choice depends on the nature of the problem and the characteristics of the dataset.             The confusion matrix is a table that is often used to evaluate the performance of a classification algorithm. It summarizes the results of a classification task, showing the number of true positive (TP), true negative (TN), false positive (FP), and false negative (FN) predictions. The confusion matrix is particularly useful for understanding the types and quantities of errors made by a classifier.

Here's how to interpret the confusion matrix you provided:

```plaintext
            Actual Positive     Actual Negative
Predicted Positive     690 (True Positive)      29 (False Positive)
Predicted Negative     73 (False Negative)      243 (True Negative)
```

- **True Positive (TP):** The number of instances correctly predicted as positive (in this case, 'Spam').
- **True Negative (TN):** The number of instances correctly predicted as negative (in this case, 'Not Spam').
- **False Positive (FP):** The number of instances incorrectly predicted as positive when they are actually negative (Type I error).
- **False Negative (FN):** The number of instances incorrectly predicted as negative when they are actually positive (Type II error).

In the provided confusion matrix:

- There are 690 instances correctly predicted as 'Spam' (True Positives).
- There are 243 instances correctly predicted as 'Not Spam' (True Negatives).
- There are 29 instances incorrectly predicted as 'Spam' when they are actually 'Not Spam' (False Positives).
- There are 73 instances incorrectly predicted as 'Not Spam' when they are actually 'Spam' (False Negatives).

These values can be used to calculate various metrics such as accuracy, precision, recall, and F1-score, providing a more comprehensive evaluation of the model's performance beyond just accuracy.